taskqueue/taskqueue.go

```diff
@@ -43,6 +46,7 @@ func NewTaskQueue(ctx context.Context) *WorkerTaskQueue {
 		cancelFn:      cancelFn,
 		peerTaskQueue: peertaskqueue.New(),
 		workSignal:    make(chan struct{}, 1),
+		noTaskSignal:  make(chan struct{}, 1),
```

Comment:
My two cents, and this might be why the method isn't working as expected for you:

* the channel only buffers one "no tasks" signal
* `worker` only tries to signal via the channel when finishing a task results in `activeTasks` being 0
* `WaitForNoActiveTasks` consumes that channel signal

So, for instance, if two `WaitForNoActiveTasks` calls come in and the worker performs one task, it will only send one signal to the channel, and only one of the two `WaitForNoActiveTasks` calls will grab that signal - the other call will potentially block forever.

You want a "broadcast signal", which unfortunately doesn't have a primitive in Go right now - channels are one-to-one. You can use `close` as a way to broadcast to infinite receivers, but that closes the channel forever, not allowing its reuse.

If you really only call `WaitForNoActiveTasks` once at a time, then your code isn't buggy, but I'd still argue it's prone to misuse :) So I would adapt its code to make it panic if called while it's already waiting.

If you do want to support concurrent Wait calls, then we'd need a different approach. Happy to bounce some ideas if that's the case.